packages/expo-notifications/ios/EXNotifications/Permissions/EXUserFacingNotificationsPermissionsRequester.m

```diff
@@ -0,0 +1,221 @@
+  __block NSNumber *allowsDisplayInNotificationCenter;
+  __block NSNumber *allowsDisplayOnLockScreen;
+  __block NSNumber *allowsDisplayInCarPlay;
+  __block NSNumber *allowsAlert;
+  __block NSNumber *allowsBadge;
+  __block NSNumber *allowsSound;
+  __block NSNumber *allowsCriticalAlerts;
+
+  __block NSNumber *alertStyle;
+  __block NSNumber *allowsPreviews;
+  __block NSNumber *providesAppNotificationSettings;
+
+  __block NSNumber *allowsAnnouncements;
+
+  [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings *settings) {
+    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
+      generalStatus = UMPermissionStatusGranted;
+    } else if (settings.authorizationStatus == UNAuthorizationStatusDenied) {
+      generalStatus = UMPermissionStatusDenied;
+    }
+
+    status = [self authorizationStatusToEnum:settings.authorizationStatus];
+
+    allowsDisplayInNotificationCenter = [self notificationSettingToNumber:settings.notificationCenterSetting];
+    allowsDisplayOnLockScreen = [self notificationSettingToNumber:settings.lockScreenSetting];
+    allowsDisplayInCarPlay = [self notificationSettingToNumber:settings.carPlaySetting];
+    allowsAlert = [self notificationSettingToNumber:settings.alertSetting];
+    allowsBadge = [self notificationSettingToNumber:settings.badgeSetting];
+    allowsSound = [self notificationSettingToNumber:settings.soundSetting];
+    if (@available(iOS 12.0, *)) {
+      allowsCriticalAlerts = [self notificationSettingToNumber:settings.criticalAlertSetting];
+    }
+
+    alertStyle = [self alertStyleToEnum:settings.alertStyle];
+    if (@available(iOS 11.0, *)) {
+      allowsPreviews = [self showPreviewsSettingToEnum:settings.showPreviewsSetting];
+    }
+    if (@available(iOS 12.0, *)) {
+      providesAppNotificationSettings = @(settings.providesAppNotificationSettings);
+    }
+
+    if (@available(iOS 13.0, *)) {
+      allowsAnnouncements = [self notificationSettingToNumber:settings.announcementSetting];
+    }
+
+    dispatch_semaphore_signal(sem);
+  }];
+
+  dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
+
+  return @{
+           @"status": @(generalStatus),
+           @"ios": @{
+                     @"status": status,
+                     @"allowsDisplayInNotificationCenter": allowsDisplayInNotificationCenter ?: [NSNull null],
+                     @"allowsDisplayOnLockScreen": allowsDisplayOnLockScreen ?: [NSNull null],
+                     @"allowsDisplayInCarPlay": allowsDisplayInCarPlay ?: [NSNull null],
+                     @"allowsAlert": allowsAlert ?: [NSNull null],
+                     @"allowsBadge": allowsBadge ?: [NSNull null],
+                     @"allowsSound": allowsSound ?: [NSNull null],
+                     @"allowsCriticalAlerts": allowsCriticalAlerts ?: [NSNull null],
+
+                     @"alertStyle": alertStyle,
+                     @"allowsPreviews": allowsPreviews,
```

Comment:
Good catch!

Reply:
I have replaced the post-block-dictionary-creation with in-block-dictionary-creation. This way the dictionary value will be able to convey a new kind of information:
- `true` — permission granted
- `false` — permission denied
- `null` — application does not support this type of permission (`UNNotificationSettingNotSupported`)
- `undefined` — the OS does not expose this permission (eg. `allowsPreviews` is supported since iOS 11)

I have also used this method when serializing `UNLocationNotificationTrigger` where some of the serialized values may be omitted, left undefined, if the OS does not expose them.